py/redrock/igm.py

```diff
@@ -0,0 +1,447 @@
+
+    Returns:
+        array of float: transmitted flux fraction T[nz, nlambda]
+    """
+    global IGM
+    if IGM is None:
+        IGM = Inoue14()
+
+    if use_gpu:
+        xp = cp
+    else:
+        xp = np
+
+    #- tile observer frame wavelengths lObs to match lRest dimensions
+    lObs = xp.tile(lObs, (zObj.size, 1))
+
+    # Transmission array to fill in
+    T = xp.ones(lObs.shape)
+
+    # Only process redshifts with restframe wavelenths at or shorter than LyA.
+    min_restwave_per_z = lObs.min()/(1.+zObj)
+    ii = min_restwave_per_z < constants.LyA_wavelength
+    T[ii, :] = IGM.full_IGM(zObj[ii], lObs[ii,:], use_gpu=use_gpu)
+
+    return T
+
+
+def transmission_Lyman_CaluraKamble(zObj,lObs, use_gpu=False,
+                                    model='Calura12'):
+    """Calculate the transmitted flux fraction from the Lyman series
+    This returns the transmitted flux fraction:
+    1 -> everything is transmitted (medium is transparent)
+    0 -> nothing is transmitted (medium is opaque)
+
+    Args:
+        zObj (array of float): Redshifts of objects
+        lObs (array of float): wavelength grid
+        use_gpu (boolean): whether to use CUPY
+        model (str): Calura12 or Kamble20, IGM model constants to use
+
+    Returns:
+        array of float: transmitted flux fraction T[nz, nlambda]
+    """
+    if use_gpu:
+        xp = cp
+    else:
+        xp = np
+
+    Lyman_series = constants.Lyman_series[model]
+
+    #- lRest[num_redshifts, num_wavelengths]
+    #- restframe wavelengths for lObs for each zObj
+    lRest = xp.outer(1.0/(1.0+zObj), lObs)
+    min_wave = lRest.min()
+
+    #- tile observer frame wavelengths lObs to match lRest dimensions
+    lObs = xp.tile(lObs, (zObj.size, 1))
+
+    #- Transmission array to fill in
+    T = xp.ones(lRest.shape)
+
+    for l in list(Lyman_series.keys()):
+        if (min_wave > Lyman_series[l]['line']):
+            continue
+
+        w      = lRest<Lyman_series[l]['line']
+        zpix   = lObs[w]/Lyman_series[l]['line']-1.
+        tauEff = Lyman_series[l]['A']*(1.+zpix)**Lyman_series[l]['B']
+        T[w]  *= np.exp(-tauEff)
+
+    return T
+
+
+igm_models = ('Calura12', 'Kamble20', 'Inoue14', 'None', None)
+
+def transmission_Lyman(zObj,lObs, use_gpu=False, model='Calura12', always_return_array=True):
+    """Calculate the transmitted flux fraction from the Lyman series
+    This returns the transmitted flux fraction:
+    1 -> everything is transmitted (medium is transparent)
+    0 -> nothing is transmitted (medium is opaque)
+
+    This method will handle 3 options:
+    1 -> For GPU mode, zObj is an array of all z for a template and the return
+    value will be a cupy array (nz x nlambda)
+    2-> In CPU mode, it can auto-detect if zObj is a numpy array and if so,
+    again, all z will be processed as a vector and the return value
+    will be a numpy array (nz x nlambda)
+    3-> For legacy, it is still supported to pass zObj as a float and
```

Comment:
This line in the docstring here is not strictly true. Scalar z only works for Calura/Kamble models, for Inoue scalar z will break due to #257. Is it worth pursuing a fix to the aforementioned issue or at the very least updating the docstring before merging?

Reply:
Thanks for catching that.  I actually had updated the function for full support of all combinations of vector vs. scalar zObj and lObs for both CPU and GPU along with unit tests to prove that it worked (8 combinations!) ... and then forgot to update the docstring!  I've corrected the docstring.

Note that the wrapper function transmission_Lyman is the only one with full support for all combos.  The underlying functions transmission_IGM_Inoue14 and transmission_Lyman_CaluraKamble now both require array inputs and produce 2D array outputs, and the wrapper function handles up/down shaping those to scalars, 1D arrays, and 2D arrays.